File: performance-comparison/large-scale-test.go
package main

import (
	"fmt"
	"math/rand"
	"runtime"
	"sync"
	"sync/atomic"
	"time"
)

// 基础类型定义
type NodeKind int

const (
	FunctionDeclaration NodeKind = iota + 1
	VariableDeclaration
	CallExpression
	BinaryExpression
	Identifier
)

type ASTNode struct {
	Kind     NodeKind
	Name     string
	Children []*ASTNode
	Parent   *ASTNode
}

type Symbol struct {
	Name  string
	Type  string
	Scope int
}

// 生成 AST 的函数
func generateAST(depth, breadth int) *ASTNode {
	kinds := []NodeKind{
		FunctionDeclaration,
		VariableDeclaration,
		CallExpression,
		BinaryExpression,
		Identifier,
	}

	node := &ASTNode{
		Kind:     kinds[rand.Intn(len(kinds))],
		Name:     fmt.Sprintf("node_%d", rand.Int()),
		Children: make([]*ASTNode, 0),
	}

	if depth > 0 {
		for i := 0; i < breadth; i++ {
			child := generateAST(depth-1, breadth)
			child.Parent = node
			node.Children = append(node.Children, child)
		}
	}

	return node
}

// 模拟大型项目的数据结构
type LargeProject struct {
	Files         []*SourceFile
	GlobalSymbols *GlobalSymbolTable
	Dependencies  map[string][]string
}

type SourceFile struct {
	Path    string
	Content []byte
	AST     *ASTNode
	Symbols []*Symbol
	Size    int
}

type GlobalSymbolTable struct {
	mu      sync.RWMutex
	symbols map[string]*Symbol
	types   map[string]*TypeInfo
}

type TypeInfo struct {
	Name       string
	Properties map[string]string
	Methods    []string
}

// 创建大型项目模拟
func createLargeProject(fileCount int) *LargeProject {
	fmt.Printf("正在创建项目结构...")
	project := &LargeProject{
		Files:         make([]*SourceFile, fileCount),
		GlobalSymbols: &GlobalSymbolTable{
			symbols: make(map[string]*Symbol),
			types:   make(map[string]*TypeInfo),
		},
		Dependencies: make(map[string][]string),
	}

	// 创建大量文件
	for i := 0; i < fileCount; i++ {
		// 显示进度
		if i%50 == 0 || i == fileCount-1 {
			progress := float64(i+1) / float64(fileCount) * 100
			fmt.Printf("\r正在创建项目结构... %.1f%% (%d/%d)", progress, i+1, fileCount)
		}
		
		file := &SourceFile{
			Path:    fmt.Sprintf("src/file_%d.ts", i),
			Content: make([]byte, 10000), // 10KB 每个文件
			AST:     generateAST(6, 3),   // 减少 AST 深度以提高速度
			Symbols: make([]*Symbol, 50), // 减少符号数量
			Size:    10000,
		}

		// 填充符号
		for j := 0; j < 50; j++ {
			symbol := &Symbol{
				Name:  fmt.Sprintf("symbol_%d_%d", i, j),
				Type:  "function",
				Scope: j / 10,
			}
			file.Symbols[j] = symbol
			
			// 添加到全局符号表
			project.GlobalSymbols.symbols[symbol.Name] = symbol
		}

		// 添加类型信息
		typeInfo := &TypeInfo{
			Name:       fmt.Sprintf("Type_%d", i),
			Properties: make(map[string]string),
			Methods:    make([]string, 10), // 减少方法数量
		}
		
		for k := 0; k < 10; k++ {
			typeInfo.Properties[fmt.Sprintf("prop_%d", k)] = "string"
			typeInfo.Methods[k] = fmt.Sprintf("method_%d", k)
		}
		
		project.GlobalSymbols.types[typeInfo.Name] = typeInfo
		project.Files[i] = file
	}
	
	fmt.Printf("\n项目创建完成！\n")
	return project
}

// 单线程处理
func processProjectSingleThread(project *LargeProject) time.Duration {
	start := time.Now()
	total := len(project.Files)
	
	for i, file := range project.Files {
		if i%20 == 0 || i == total-1 {
			progress := float64(i+1) / float64(total) * 100
			fmt.Printf("\r  单线程处理进度: %.1f%% (%d/%d)", progress, i+1, total)
		}
		processFile(file, project.GlobalSymbols)
	}
	fmt.Printf("\n")
	
	return time.Since(start)
}

// 并发处理
func processProjectConcurrent(project *LargeProject) time.Duration {
	start := time.Now()
	total := len(project.Files)
	processed := int64(0)
	
	var wg sync.WaitGroup
	semaphore := make(chan struct{}, runtime.NumCPU())
	
	// 启动进度监控
	done := make(chan bool)
	go func() {
		ticker := time.NewTicker(100 * time.Millisecond)
		defer ticker.Stop()
		
		for {
			select {
			case <-ticker.C:
				current := processed
				progress := float64(current) / float64(total) * 100
				fmt.Printf("\r  并发处理进度: %.1f%% (%d/%d)", progress, current, total)
			case <-done:
				return
			}
		}
	}()
	
	for _, file := range project.Files {
		wg.Add(1)
		go func(f *SourceFile) {
			defer wg.Done()
			semaphore <- struct{}{}
			defer func() { <-semaphore }()
			
			processFile(f, project.GlobalSymbols)
			atomic.AddInt64(&processed, 1)
		}(file)
	}
	
	wg.Wait()
	done <- true
	fmt.Printf("\n")
	
	return time.Since(start)
}

// 高并发处理（模拟真实编译器）
func processProjectHighConcurrency(project *LargeProject) time.Duration {
	start := time.Now()
	total := len(project.Files)
	processed := int64(0)
	
	var wg sync.WaitGroup
	// 使用更多的 goroutine
	semaphore := make(chan struct{}, runtime.NumCPU()*4)
	
	// 启动进度监控
	done := make(chan bool)
	go func() {
		ticker := time.NewTicker(100 * time.Millisecond)
		defer ticker.Stop()
		
		for {
			select {
			case <-ticker.C:
				current := processed
				progress := float64(current) / float64(total) * 100
				fmt.Printf("\r  高并发处理进度: %.1f%% (%d/%d)", progress, current, total)
			case <-done:
				return
			}
		}
	}()
	
	for _, file := range project.Files {
		wg.Add(1)
		go func(f *SourceFile) {
			defer wg.Done()
			semaphore <- struct{}{}
			defer func() { <-semaphore }()
			
			// 更复杂的处理
			processFileWithDependencies(f, project.GlobalSymbols, project.Dependencies)
			atomic.AddInt64(&processed, 1)
		}(file)
	}
	
	wg.Wait()
	done <- true
	fmt.Printf("\n")
	
	return time.Since(start)
}

// 模拟文件处理
func processFile(file *SourceFile, globalSymbols *GlobalSymbolTable) {
	// 模拟 AST 遍历
	nodeCount := visitNodeComplex(file.AST)
	
	// 模拟符号解析
	for _, symbol := range file.Symbols {
		resolveSymbolWithGlobal(symbol.Name, globalSymbols)
	}
	
	// 模拟类型检查
	for i := 0; i < nodeCount/10; i++ {
		performTypeCheck(globalSymbols)
	}
}

// 带依赖关系的文件处理
func processFileWithDependencies(file *SourceFile, globalSymbols *GlobalSymbolTable, deps map[string][]string) {
	// 基本处理
	processFile(file, globalSymbols)
	
	// 处理依赖关系
	if dependencies, exists := deps[file.Path]; exists {
		for _, dep := range dependencies {
			// 模拟依赖解析
			resolveSymbolWithGlobal(dep, globalSymbols)
		}
	}
	
	// 模拟增量编译检查
	checkIncrementalChanges(file, globalSymbols)
}

// 复杂的 AST 遍历
func visitNodeComplex(node *ASTNode) int {
	if node == nil {
		return 0
	}
	
	count := 1
	
	// 减少模拟操作以提高速度
	for i := 0; i < 50; i++ {
		_ = fmt.Sprintf("op_%s_%d", node.Name, i)
	}
	
	// 递归处理子节点
	for _, child := range node.Children {
		count += visitNodeComplex(child)
	}
	
	return count
}

// 全局符号解析
func resolveSymbolWithGlobal(name string, globalSymbols *GlobalSymbolTable) *Symbol {
	globalSymbols.mu.RLock()
	defer globalSymbols.mu.RUnlock()
	
	if symbol, exists := globalSymbols.symbols[name]; exists {
		return symbol
	}
	
	return nil
}

// 类型检查
func performTypeCheck(globalSymbols *GlobalSymbolTable) {
	globalSymbols.mu.RLock()
	defer globalSymbols.mu.RUnlock()
	
	// 简化类型推导
	count := 0
	for _, typeInfo := range globalSymbols.types {
		for range typeInfo.Properties {
			count++
			if count > 5 { // 限制处理数量
				return
			}
		}
	}
}

// 增量编译检查
func checkIncrementalChanges(file *SourceFile, globalSymbols *GlobalSymbolTable) {
	// 简化增量编译逻辑
	for i := 0; i < 10; i++ {
		symbolName := fmt.Sprintf("inc_%d", i)
		resolveSymbolWithGlobal(symbolName, globalSymbols)
	}
}

// 内存使用统计
func getMemStats() (float64, float64) {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	return float64(m.Alloc) / 1024 / 1024, float64(m.Sys) / 1024 / 1024
}

func main() {
	fmt.Println("=== 大规模 Go 并发测试 ===\n")
	
	// 调整测试规模，使其更合理
	fileCounts := []int{50, 200, 500}
	
	for _, fileCount := range fileCounts {
		fmt.Printf("测试项目规模: %d 个文件\n", fileCount)
		fmt.Println("----------------------------------------")
		
		// 创建项目
		project := createLargeProject(fileCount)
		
		allocBefore, _ := getMemStats()
		
		// 单线程测试
		fmt.Println("单线程处理...")
		singleTime := processProjectSingleThread(project)
		
		// 并发测试
		fmt.Println("并发处理...")
		concurrentTime := processProjectConcurrent(project)
		
		// 高并发测试
		fmt.Println("高并发处理...")
		highConcurrentTime := processProjectHighConcurrency(project)
		
		allocAfter, _ := getMemStats()
		
		// 结果
		fmt.Printf("\n结果:\n")
		fmt.Printf("  单线程耗时: %.2f ms\n", float64(singleTime.Nanoseconds())/1000000)
		fmt.Printf("  并发耗时: %.2f ms\n", float64(concurrentTime.Nanoseconds())/1000000)
		fmt.Printf("  高并发耗时: %.2f ms\n", float64(highConcurrentTime.Nanoseconds())/1000000)
		fmt.Printf("  并发提升: %.2fx\n", float64(singleTime.Nanoseconds())/float64(concurrentTime.Nanoseconds()))
		fmt.Printf("  高并发提升: %.2fx\n", float64(singleTime.Nanoseconds())/float64(highConcurrentTime.Nanoseconds()))
		fmt.Printf("  内存使用: %.2f MB\n", allocAfter-allocBefore)
		fmt.Printf("  CPU 核心数: %d\n\n", runtime.NumCPU())
	}

} 

File: performance-comparison/src/large-scale-test.ts
// TypeScript 大规模性能测试

// 基础类型定义
enum LargeTestNodeKind {
  FunctionDeclaration = 1,
  VariableDeclaration = 2,
  CallExpression = 3,
  BinaryExpression = 4,
  Identifier = 5
}

class LargeTestASTNode {
  kind: LargeTestNodeKind;
  name: string;
  children: LargeTestASTNode[];
  parent: LargeTestASTNode | undefined;

  constructor(kind: LargeTestNodeKind, name: string) {
    this.kind = kind;
    this.name = name;
    this.children = [];
    this.parent = undefined;
  }
}

class TSSymbol {
  name: string;
  type: string;
  scope: number;

  constructor(name: string, type: string, scope: number) {
    this.name = name;
    this.type = type;
    this.scope = scope;
  }
}

// 生成 AST 的函数
function generateAST(depth: number, breadth: number): LargeTestASTNode {
  const kinds: LargeTestNodeKind[] = [
    LargeTestNodeKind.FunctionDeclaration,
    LargeTestNodeKind.VariableDeclaration,
    LargeTestNodeKind.CallExpression,
    LargeTestNodeKind.BinaryExpression,
    LargeTestNodeKind.Identifier,
  ];

  const node = new LargeTestASTNode(
    kinds[Math.floor(Math.random() * kinds.length)],
    `node_${Math.floor(Math.random() * 1000000)}`
  );

  if (depth > 0) {
    for (let i = 0; i < breadth; i++) {
      const child = generateAST(depth - 1, breadth);
      child.parent = node;
      node.children.push(child);
    }
  }

  return node;
}

// 模拟大型项目的数据结构
class LargeProject {
  files: SourceFile[];
  globalSymbols: GlobalSymbolTable;
  dependencies: Map<string, string[]>;

  constructor() {
    this.files = [];
    this.globalSymbols = new GlobalSymbolTable();
    this.dependencies = new Map();
  }
}

class SourceFile {
  path: string;
  content: number[];
  ast: LargeTestASTNode;
  symbols: TSSymbol[];
  size: number;

  constructor(path: string, size: number) {
    this.path = path;
    this.content = new Array(size).fill(0); // 模拟文件内容
    this.ast = generateAST(6, 3); // 减少 AST 深度以提高速度
    this.symbols = [];
    this.size = size;
  }
}

class GlobalSymbolTable {
  private symbols: Map<string, TSSymbol>;
  private types: Map<string, TypeInfo>;

  constructor() {
    this.symbols = new Map();
    this.types = new Map();
  }

  addSymbol(symbol: TSSymbol): void {
    this.symbols.set(symbol.name, symbol);
  }

  getSymbol(name: string): TSSymbol | undefined {
    return this.symbols.get(name);
  }

  addType(typeInfo: TypeInfo): void {
    this.types.set(typeInfo.name, typeInfo);
  }

  getType(name: string): TypeInfo | undefined {
    return this.types.get(name);
  }

  get typesIterator(): IterableIterator<[string, TypeInfo]> {
    return this.types.entries();
  }
}

class TypeInfo {
  name: string;
  properties: Map<string, string>;
  methods: string[];

  constructor(name: string) {
    this.name = name;
    this.properties = new Map();
    this.methods = [];
  }
}

// 创建大型项目模拟
function createLargeProject(fileCount: number): LargeProject {
  process.stdout.write('正在创建项目结构...');
  const project = new LargeProject();

  // 创建大量文件
  for (let i = 0; i < fileCount; i++) {
    // 显示进度
    if (i % 50 === 0 || i === fileCount - 1) {
      const progress = ((i + 1) / fileCount * 100).toFixed(1);
      process.stdout.write(`\r正在创建项目结构... ${progress}% (${i + 1}/${fileCount})`);
    }

    const file = new SourceFile(`src/file_${i}.ts`, 10000); // 10KB 每个文件

    // 填充符号
    for (let j = 0; j < 50; j++) {
      const symbol = new TSSymbol(
        `symbol_${i}_${j}`,
        'function',
        Math.floor(j / 10)
      );
      file.symbols.push(symbol);

      // 添加到全局符号表
      project.globalSymbols.addSymbol(symbol);
    }

    // 添加类型信息
    const typeInfo = new TypeInfo(`Type_${i}`);

    for (let k = 0; k < 10; k++) {
      typeInfo.properties.set(`prop_${k}`, 'string');
      typeInfo.methods.push(`method_${k}`);
    }

    project.globalSymbols.addType(typeInfo);
    project.files.push(file);
  }

  console.log('\n项目创建完成！');
  return project;
}

// 单线程处理
function processProjectSingleThread(project: LargeProject): number {
  const start = process.hrtime.bigint();
  const total = project.files.length;

  for (let i = 0; i < project.files.length; i++) {
    if (i % 20 === 0 || i === total - 1) {
      const progress = ((i + 1) / total * 100).toFixed(1);
      process.stdout.write(`\r  单线程处理进度: ${progress}% (${i + 1}/${total})`);
    }
    processFile(project.files[i], project.globalSymbols);
  }
  console.log('');

  const end = process.hrtime.bigint();
  return Number(end - start) / 1000000; // 转换为毫秒
}

// 模拟并发处理（TypeScript 单线程，使用 Promise 模拟）
async function processProjectConcurrent(project: LargeProject): Promise<number> {
  const start = process.hrtime.bigint();
  const total = project.files.length;
  let processed = 0;

  // 启动进度监控
  const progressInterval = setInterval(() => {
    const progress = (processed / total * 100).toFixed(1);
    process.stdout.write(`\r  模拟并发处理进度: ${progress}% (${processed}/${total})`);
  }, 100);

  // 分批处理文件（模拟并发）
  const batchSize = Math.max(1, Math.floor(total / 10));
  const batches: SourceFile[][] = [];

  for (let i = 0; i < project.files.length; i += batchSize) {
    const batch = project.files.slice(i, i + batchSize);
    batches.push(batch);
  }

  // 处理每个批次
  for (const batch of batches) {
    const promises = batch.map((file: SourceFile) => {
      return new Promise<void>(resolve => {
        // 使用 setImmediate 模拟异步处理
        setImmediate(() => {
          processFile(file, project.globalSymbols);
          processed++;
          resolve();
        });
      });
    });

    await Promise.all(promises);
  }

  clearInterval(progressInterval);
  console.log('');

  const end = process.hrtime.bigint();
  return Number(end - start) / 1000000; // 转换为毫秒
}

// 高并发处理（使用 Worker Threads 模拟，这里简化为更小的批次）
async function processProjectHighConcurrency(project: LargeProject): Promise<number> {
  const start = process.hrtime.bigint();
  const total = project.files.length;
  let processed = 0;

  // 启动进度监控
  const progressInterval = setInterval(() => {
    const progress = (processed / total * 100).toFixed(1);
    process.stdout.write(`\r  高并发模拟处理进度: ${progress}% (${processed}/${total})`);
  }, 100);

  // 更小的批次，模拟更高的并发
  const batchSize = Math.max(1, Math.floor(total / 20));
  const batches: SourceFile[][] = [];

  for (let i = 0; i < project.files.length; i += batchSize) {
    const batch = project.files.slice(i, i + batchSize);
    batches.push(batch);
  }

  // 处理每个批次
  for (const batch of batches) {
    const promises = batch.map((file: SourceFile) => {
      return new Promise<void>(resolve => {
        setImmediate(() => {
          // 更复杂的处理
          processFileWithDependencies(file, project.globalSymbols, project.dependencies);
          processed++;
          resolve();
        });
      });
    });

    await Promise.all(promises);
  }

  clearInterval(progressInterval);
  console.log('');

  const end = process.hrtime.bigint();
  return Number(end - start) / 1000000; // 转换为毫秒
}

// 模拟文件处理
function processFile(file: SourceFile, globalSymbols: GlobalSymbolTable): void {
  // 模拟 AST 遍历
  const nodeCount = visitNodeComplex(file.ast);

  // 模拟符号解析
  for (const symbol of file.symbols) {
    resolveSymbolWithGlobal(symbol.name, globalSymbols);
  }

  // 模拟类型检查
  for (let i = 0; i < Math.floor(nodeCount / 10); i++) {
    performTypeCheck(globalSymbols);
  }
}

// 带依赖关系的文件处理
function processFileWithDependencies(
  file: SourceFile, 
  globalSymbols: GlobalSymbolTable, 
  dependencies: Map<string, string[]>
): void {
  // 基本处理
  processFile(file, globalSymbols);

  // 处理依赖关系
  if (dependencies.has(file.path)) {
    const deps = dependencies.get(file.path)!;
    for (const dep of deps) {
      // 模拟依赖解析
      resolveSymbolWithGlobal(dep, globalSymbols);
    }
  }

  // 模拟增量编译检查
  checkIncrementalChanges(file, globalSymbols);
}

// 复杂的 AST 遍历
function visitNodeComplex(node: LargeTestASTNode | undefined): number {
  if (!node) {
    return 0;
  }

  let count = 1;

  // 减少模拟操作以提高速度
  for (let i = 0; i < 50; i++) {
    const temp = `op_${node.name}_${i}`;
  }

  // 递归处理子节点
  for (const child of node.children) {
    count += visitNodeComplex(child);
  }

  return count;
}

// 全局符号解析
function resolveSymbolWithGlobal(name: string, globalSymbols: GlobalSymbolTable): TSSymbol | undefined {
  return globalSymbols.getSymbol(name);
}

// 类型检查
function performTypeCheck(globalSymbols: GlobalSymbolTable): void {
  // 简化类型推导
  let count = 0;
  for (const [, typeInfo] of globalSymbols.typesIterator) {
    for (const [,] of typeInfo.properties) {
      count++;
      if (count > 5) { // 限制处理数量
        return;
      }
    }
  }
}

// 增量编译检查
function checkIncrementalChanges(file: SourceFile, globalSymbols: GlobalSymbolTable): void {
  // 简化增量编译逻辑
  for (let i = 0; i < 10; i++) {
    const symbolName = `inc_${i}`;
    resolveSymbolWithGlobal(symbolName, globalSymbols);
  }
}

// 内存使用统计
interface MemStats {
  heapUsed: number;
  heapTotal: number;
}

function getMemStats(): MemStats {
  const memUsage = process.memoryUsage();
  return {
    heapUsed: memUsage.heapUsed / 1024 / 1024,
    heapTotal: memUsage.heapTotal / 1024 / 1024
  };
}

// 主函数
async function main(): Promise<void> {
  console.log('=== 大规模 TypeScript 测试 ===\n');

  // 调整测试规模，使其更合理
  const fileCounts: number[] = [50, 200, 500];

  for (const fileCount of fileCounts) {
    console.log(`测试项目规模: ${fileCount} 个文件`);
    console.log('----------------------------------------');

    // 创建项目
    const project = createLargeProject(fileCount);

    const memBefore = getMemStats();

    // 单线程测试
    console.log('单线程处理...');
    const singleTime = processProjectSingleThread(project);

    // 模拟并发测试
    console.log('模拟并发处理...');
    const concurrentTime = await processProjectConcurrent(project);

    // 高并发测试
    console.log('高并发模拟处理...');
    const highConcurrentTime = await processProjectHighConcurrency(project);

    const memAfter = getMemStats();

    // 结果
    console.log('\n结果:');
    console.log(`  单线程耗时: ${singleTime.toFixed(2)} ms`);
    console.log(`  模拟并发耗时: ${concurrentTime.toFixed(2)} ms`);
    console.log(`  高并发模拟耗时: ${highConcurrentTime.toFixed(2)} ms`);
    console.log(`  模拟并发提升: ${(singleTime / concurrentTime).toFixed(2)}x`);
    console.log(`  高并发模拟提升: ${(singleTime / highConcurrentTime).toFixed(2)}x`);
    console.log(`  内存使用: ${(memAfter.heapUsed - memBefore.heapUsed).toFixed(2)} MB`);
    console.log(`  CPU 核心数: ${require('os').cpus().length}\n`);
  }
}

// 运行测试
main().catch(console.error); 